src/Hacknet/HacknetHelpers.tsx

```diff
@@ -460,139 +463,137 @@ export function updateHashManagerCapacity(): void {
-        break;
-      }
-      case "Generate Coding Contract": {
+        //Manually loop the change so as to properly handle the softcap
         for (let i = 0; i < count; i++) {
-          generateRandomContract();
+          target.changeMaximumMoney(upg.value);
         }
-        break;
+      } catch (e) {
+        Player.hashManager.refundUpgrade(upgName, count);
+        return { success: false, message: e instanceof Error ? e.message : String(e) };
       }
-      case "Company Favor": {
-        if (!isMember("CompanyName", upgTarget)) {
-          console.error(`Invalid target specified in purchaseHashUpgrade(): ${upgTarget}`);
-          throw new Error(`'${upgTarget}' is not a company.`);
-        }
-        Companies[upgTarget].setFavor(Companies[upgTarget].favor + 5 * count);
-        break;
+      break;
+    }
+    case HashUpgradeEnum.ImproveStudying: {
+      // Multiplier is handled by HashManager
+      break;
+    }
+    case HashUpgradeEnum.ImproveGymTraining: {
+      // Multiplier is handled by HashManager
+      break;
+    }
+    case HashUpgradeEnum.ExchangeForCorporationResearch: {
+      const corp = Player.corporation;
+      if (corp === null) {
+        Player.hashManager.refundUpgrade(upgName, count);
+        return { success: false, message: "You have not created a corporation." };
+      }
+      for (const division of corp.divisions.values()) {
+        division.researchPoints += upg.value * count;
+      }
+      break;
+    }
+    case HashUpgradeEnum.ExchangeForBladeburnerRank: {
+      const bladeburner = Player.bladeburner;
+      if (bladeburner === null) {
+        Player.hashManager.refundUpgrade(upgName, count);
+        return { success: false, message: "You have not joined Bladeburner." };
       }
-      default:
-        console.warn(`Unrecognized upgrade name ${upgName}. Upgrade has no effect`);
-        return false;
+      bladeburner.changeRank(Player, upg.value * count);
+      break;
     }
+    case HashUpgradeEnum.ExchangeForBladeburnerSP: {
+      const bladeburner = Player.bladeburner;
+      if (bladeburner === null) {
+        Player.hashManager.refundUpgrade(upgName, count);
+        return { success: false, message: "You have not joined Bladeburner." };
+      }
 
-    return true;
+      bladeburner.skillPoints += upg.value * count;
+      break;
+    }
+    case HashUpgradeEnum.GenerateCodingContract: {
+      for (let i = 0; i < count; i++) {
+        generateRandomContract();
+      }
+      break;
+    }
+    case HashUpgradeEnum.CompanyFavor: {
+      if (!isMember("CompanyName", upgTarget)) {
+        return { success: false, message: `'${upgTarget}' is not a company.` };
+        // throw new Error(`'${upgTarget}' is not a company.`);
+      }
+      Companies[upgTarget].setFavor(Companies[upgTarget].favor + 5 * count);
+      break;
+    }
+    default:
+      return { success: false, message: `Unrecognized upgrade name ${upgName}. Upgrade has no effect` };
```

Comment:
You can only reach here if hashManager.upgrade explicitly returns success, so it seems odd to then return failure. I think we want the "throw unreachable" construct here, or ideally get TS to tell us all cases are covered.